test/plugins/test-trace-http.ts

```diff
@@ -13,357 +13,420 @@
+  // this.handleDone has been called.
+  done: Promise<string>;
+  // A callback to be passed to http.request or http.get, so that when response
+  // data has been fully consumed, this.done will be resolved.
+  handleResponse = (res: httpModule.IncomingMessage) => {
+    let data = '';
+    res.on('data', d => data += d);
+    res.on('error', this.reject);
+    res.on('end', () => this.resolve(data));
+  };
 
-    server = http.Server(function(req, res) {
-      setTimeout(function() {
-        res.writeHead(200);
-        res.end(common.serverRes);
-      }, common.serverWait);
+  constructor() {
+    this.done = new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
     });
-  });
+  }
 
-  afterEach(function() {
-    common.cleanTraces();
-    server.close();
-  });
+  /**
+   * Resolves this.done. This should be used when passing handleResponse
+   * to a function isn't applicable.
+   */
+  handleDone() {
+    this.resolve('');
+  }
+}
+
+const wait = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
 
-  it('should patch the necessary functions', function() {
-    assert.strictEqual(http.request.__wrapped, true);
-    if (semver.satisfies(process.version, '>=8.0.0')) {
-      assert.strictEqual(http.get.__wrapped, true, 'should patch get');
-    } else {
-      assert.strictEqual(http.get.__wrapped, undefined, 'should not patch get');
+// Server abstraction class definitions. These are borrowed from web framework
+// tests -- which are useful because they already expose a Promise API.
+const servers = {
+  http: Express4, https: class Express4Secure extends Express4 {
+    static key = fs.readFileSync(path.join(__dirname, 'fixtures', 'key.pem'));
+    static cert = fs.readFileSync(path.join(__dirname, 'fixtures', 'cert.pem'));
+    private https: typeof httpsModule;
+
+    constructor() {
+      super();
+      this.https = require('https');
     }
-  });
 
-  it('should accurately measure get time with callback', function(done) {
-    server.listen(common.serverPort, common.runInTransaction.bind(null,
-      function(endTransaction) {
-        var start = Date.now();
-        http.get({port: common.serverPort}, function(res) {
-          var result = '';
-          res.on('data', function(data) { result += data; });
-          res.on('end', function() {
-            endTransaction();
-            assert.equal(result, common.serverRes);
-            common.assertDurationCorrect(Date.now() - start);
-            done();
-          });
-        });
-      })
-    );
-  });
+    listen(port: number): number {
+      // The types of (http|https).Server are not compatible, but we don't
+      // access any properties that aren't present on both in the test.
+      // tslint:disable:no-any
+      this.server = this.https.createServer(
+                        {key: Express4Secure.key, cert: Express4Secure.cert},
+                        this.app) as any as httpModule.Server;
```

Comment:
You can say `as {} as httpModule.Server` and delete the `tslint` pragma above.

Reply:
Done